chalk-solve/src/clauses/dyn_ty.rs

```diff
@@ -0,0 +1,181 @@
+        for exists_qwc in dyn_ty.bounds.map_ref(|r| r.iter(interner)) {
+            // Replace the `T` from `exists<T> { .. }` with `self_ty`,
+            // yielding clases like
+            //
+            // ```
+            // forall<'a> { Implemented(dyn Fn(&u8): Fn<(&'a u8)>) }
+            // ```
+            let qwc = exists_qwc.substitute(interner, &[self_ty.clone().cast(interner)]);
+
+            builder.push_binders(&qwc, |builder, wc| match &wc {
+                // For the implemented traits, we need to elaborate super traits and add where clauses from the trait
+                WhereClause::Implemented(trait_ref) => {
+                    push_dyn_ty_impl_clauses(db, builder, trait_ref.clone())
+                }
+                // Associated item bindings are just taken as facts (?)
+                WhereClause::AliasEq(_) => builder.push_fact(wc),
+            });
+        }
+    });
+}
+
+/// Generate `Implemented` clauses for a `dyn Trait` type. We need to generate
+/// `Implemented` clauses for all super traits, and for each trait we require
+/// its where clauses. (See #203.)
+fn push_dyn_ty_impl_clauses<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    trait_ref: TraitRef<I>,
+) {
+    let interner = db.interner();
+    // We have some `dyn Trait`, and some `trait SuperTrait: WC`
+    // which is a super trait of `Trait` (including actually
+    // just being the same trait); then we want to push
+    // `Implemented(dyn Trait: SuperTrait) :- WC`.
+
+    let super_trait_refs =
+        super_traits(db, trait_ref.trait_id).substitute(interner, &trait_ref.substitution);
+
+    for q_super_trait_ref in super_trait_refs {
+        builder.push_binders(&q_super_trait_ref, |builder, super_trait_ref| {
+            let trait_datum = db.trait_datum(super_trait_ref.trait_id);
+            let wc = trait_datum
+                .where_clauses()
+                .substitute(interner, &super_trait_ref.substitution);
+            builder.push_clause(super_trait_ref, wc);
+        });
+    }
+}
+
+pub fn super_traits<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    trait_id: TraitId<I>,
+) -> Binders<Vec<Binders<TraitRef<I>>>> {
+    let interner = db.interner();
+    let mut seen_traits = FxHashSet::default();
+    let trait_datum = db.trait_datum(trait_id);
+    let trait_ref = Binders::empty(
+        db.interner(),
+        TraitRef {
+            trait_id,
+            substitution: trait_datum
+                .binders
+                .identity_substitution(interner)
+                .shifted_in(interner),
+        },
+    );
+    let mut trait_refs = Vec::new();
+    go(db, trait_ref, &mut seen_traits, &mut trait_refs);
+
+    fn go<I: Interner>(
+        db: &dyn RustIrDatabase<I>,
+        trait_ref: Binders<TraitRef<I>>,
+        seen_traits: &mut FxHashSet<TraitId<I>>,
+        trait_refs: &mut Vec<Binders<TraitRef<I>>>,
+    ) {
+        let interner = db.interner();
```

Comment:
(this doesn't really have to be changed...just a passing thought)

Reply:
Yeah, I mostly put it down below because that avoids collecting and substituting those traits, but that's probably not a very important consideration...

Reply:
Meh, efficiency is a good reason. =